File: frontend/src/api/community.ts
import apiClient from './client'

export enum PostType {
  QUESTION = 'question',
  SHARE = 'share'
}

export interface UserBasicInfo {
  id: number
  name: string | null
  email: string
}

export interface PostPhoto {
  id: number
  post_id: number
  file_key: string
  display_order: number
  photo_url: string
  created_at: string
}

export interface Post {
  id: number
  user_id: number
  user: UserBasicInfo
  post_type: PostType
  content: string
  photos: PostPhoto[]
  like_count: number
  comment_count: number
  is_liked: boolean
  is_deleted: boolean
  created_at: string
  updated_at: string
}

export interface PostListResponse {
  total: number
  posts: Post[]
  hasMore: boolean
}

export interface Comment {
  id: number
  post_id: number
  user_id: number
  user: UserBasicInfo
  content: string
  like_count: number
  is_liked: boolean
  is_deleted: boolean
  created_at: string
}

export interface CommentListResponse {
  total: number
  comments: Comment[]
}

export interface LikeResponse {
  success: boolean
  isLiked: boolean
  likeCount: number
}

export async function getPosts(
  skip?: number,
  limit?: number,
  postType?: PostType,
  search?: string
): Promise<PostListResponse> {
  const params: any = {}
  if (skip !== undefined) params.skip = skip
  if (limit !== undefined) params.limit = limit
  if (postType) params.post_type = postType
  if (search) params.search = search
  
  const response = await apiClient.get('/community/posts', { params })
  return {
    total: response.data.total,
    posts: response.data.posts,
    hasMore: response.data.has_more
  }
}

export async function getMyPosts(
  skip?: number,
  limit?: number
): Promise<PostListResponse> {
  const params: any = {}
  if (skip !== undefined) params.skip = skip
  if (limit !== undefined) params.limit = limit
  
  const response = await apiClient.get('/community/posts/my', { params })
  return {
    total: response.data.total,
    posts: response.data.posts,
    hasMore: response.data.has_more
  }
}

export async function getPost(postId: number): Promise<Post> {
  const response = await apiClient.get(`/community/posts/${postId}`)
  return response.data
}

export async function createPost(formData: FormData): Promise<Post> {
  const response = await apiClient.post('/community/posts', formData, {
    headers: {
      'Content-Type': 'multipart/form-data'
    }
  })
  return response.data
}

export async function updatePost(
  postId: number, 
  content: string, 
  postType?: PostType,
  deletePhotoIds?: number[],
  newPhotos?: File[]
): Promise<Post> {
  const formData = new FormData()
  formData.append('content', content)
  if (postType) {
    formData.append('post_type', postType)
  }
  if (deletePhotoIds && deletePhotoIds.length > 0) {
    formData.append('delete_photo_ids', JSON.stringify(deletePhotoIds))
  }
  if (newPhotos && newPhotos.length > 0) {
    newPhotos.forEach(photo => {
      formData.append('photos', photo)
    })
  }
  
  const response = await apiClient.put(`/community/posts/${postId}`, formData, {
    headers: {
      'Content-Type': 'multipart/form-data'
    }
  })
  return response.data
}

export async function deletePost(postId: number): Promise<void> {
  await apiClient.delete(`/community/posts/${postId}`)
}

export async function getComments(postId: number): Promise<CommentListResponse> {
  const response = await apiClient.get(`/community/posts/${postId}/comments`)
  return response.data
}

export async function createComment(postId: number, content: string): Promise<Comment> {
  const response = await apiClient.post(`/community/posts/${postId}/comments`, {
    content
  })
  return response.data
}

export async function deleteComment(
  postId: number,
  commentId: number
): Promise<void> {
  await apiClient.delete(`/community/comments/${commentId}`)
}

export async function togglePostLike(postId: number, isCurrentlyLiked: boolean): Promise<LikeResponse> {
  const response = isCurrentlyLiked 
    ? await apiClient.delete(`/community/posts/${postId}/like`)
    : await apiClient.post(`/community/posts/${postId}/like`)
  return {
    success: response.data.success,
    isLiked: response.data.is_liked,
    likeCount: response.data.like_count
  }
}

export async function toggleCommentLike(commentId: number): Promise<LikeResponse> {
  const response = await apiClient.post(`/community/comments/${commentId}/like`)
  return {
    success: response.data.success,
    isLiked: response.data.is_liked,
    likeCount: response.data.like_count
  }
}

export async function reportPost(postId: number, reason: string): Promise<{ success: boolean; message: string }> {
  const response = await apiClient.post(`/community/posts/${postId}/report`, { reason })
  return response.data
}

File: backend/app/schemas/community.py
"""
Community Schemas
社群功能的 Pydantic schemas
"""
from pydantic import BaseModel, Field, validator
from typing import List, Optional
from datetime import datetime
from enum import Enum


class PostTypeEnum(str, Enum):
    """貼文類型"""
    question = "question"
    share = "share"


# ===== Photo Schemas =====
class PostPhotoBase(BaseModel):
    """貼文照片基礎 schema"""
    file_key: str
    display_order: int = 0


class PostPhotoResponse(PostPhotoBase):
    """貼文照片回應 schema"""
    id: int
    post_id: int
    photo_url: str  # presigned URL
    created_at: datetime

    class Config:
        from_attributes = True


# ===== Post Schemas =====
class PostCreate(BaseModel):
    """建立貼文 schema"""
    post_type: PostTypeEnum
    content: str = Field(..., min_length=1, max_length=5000)
    # photos will be uploaded separately via multipart/form-data


class PostUpdate(BaseModel):
    """更新貼文 schema"""
    content: Optional[str] = Field(None, min_length=1, max_length=5000)
    post_type: Optional[PostTypeEnum] = None


class UserBasicInfo(BaseModel):
    """用戶基本資訊"""
    id: int
    name: Optional[str]
    email: str

    class Config:
        from_attributes = True


class PostResponse(BaseModel):
    """貼文回應 schema"""
    id: int
    user_id: int
    user: UserBasicInfo
    post_type: PostTypeEnum
    content: str
    photos: List[PostPhotoResponse] = []
    like_count: int = 0
    comment_count: int = 0
    is_liked: bool = False  # 當前用戶是否已按讚
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PostListResponse(BaseModel):
    """貼文列表回應"""
    total: int
    posts: List[PostResponse]
    has_more: bool = False


# ===== Comment Schemas =====
class CommentCreate(BaseModel):
    """建立留言 schema"""
    content: str = Field(..., min_length=1, max_length=1000)


class CommentUpdate(BaseModel):
    """更新留言 schema"""
    content: str = Field(..., min_length=1, max_length=1000)


class CommentResponse(BaseModel):
    """留言回應 schema"""
    id: int
    post_id: int
    user_id: int
    user: UserBasicInfo
    content: str
    like_count: int = 0
    is_liked: bool = False  # 當前用戶是否已按讚
    is_deleted: bool = False
    created_at: datetime

    class Config:
        from_attributes = True


class CommentListResponse(BaseModel):
    """留言列表回應"""
    total: int
    comments: List[CommentResponse]


# ===== Like Schemas =====
class LikeResponse(BaseModel):
    """按讚回應 schema"""
    success: bool
    is_liked: bool  # 操作後的狀態
    like_count: int  # 總讚數

    class Config:
        from_attributes = True


# ===== Report Schemas =====
class PostReportCreate(BaseModel):
    """檢舉貼文 schema"""
    reason: str = Field(..., min_length=1, max_length=1000, description="檢舉原因")


class PostReportResponse(BaseModel):
    """檢舉貼文回應 schema"""
    success: bool
    message: str

    class Config:
        from_attributes = True
